idaes/core/util/performance.py

```diff
@@ -0,0 +1,169 @@
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES), and is copyright (c) 2018-2021
+# by the software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+"""
+Author: Andrew Lee
+"""
+import gc
+import pytest
+
+from pyomo.util.check_units import assert_units_consistent
+from pyomo.common.timing import TicTocTimer
+from pyomo.environ import assert_optimal_termination
+import pyomo.common.unittest as unittest
+
+from idaes.core.solvers import get_solver
+from idaes.core.util.model_statistics import degrees_of_freedom
+
+# Get default solver for testing
+solver = get_solver()
+
+
+class PerformanceBaseClass:
+    """
+    Base class for running the IDAES Performance Test Suite.
+
+    This class is intended to be used as a basis for constructing tests to be run as
+    part of the IDAES performance testing suite, and should implement the following
+    methods:
+
+    * build_model - should return a complete instance of the model being tested.
+    * test_performance - a test method for profiling the model of interest which logs
+        some performance data. A standard method is provided which logs some common
+        performance data, but developers should overload this as required.
+    * initialize_model - (optional) a method to initialize the model of interest. This
+        method is called as part of the standard test_performance method.
+    * solve_model - (optional) a method to solve the model of interest. This method is
+        called as part of the standard test_performance method, and a standard method
+        is provided by the base class.
+
+    Developers may add additional test methods to the derived class to run additional
+    performance tests if desired. These methods should use standard pytest marks and
+    naming conventions.
+
+    Usage:
+
+    Performance tests should inherit from this based class and unittest.TestCase.
+    Inheritance from TestCase is used to make the derived classes easily discoverable
+    for use outside the performance testing infrastructure. The derived class should
+    be decorated with the 'performance' pytest mark. This will automatically categorize
+    all tests in this class as performance tests that will only be run when explicitly
+    called for.
+
+    Raises:
+        TypeError is derived class does not inherit from unittest.TestCase.
+    """
+
+    def __init_subclass__(cls):
+        if unittest.TestCase not in cls.mro():
```

Comment:
Minor nitpick: I think `if not issubclass(cls, unittest.TestCase):` is cleaner than directly checking the MRO. Let me know if there is any technical reason to check the MRO instead.

Reply:
The reason for this is that when @jsiirola suggested this approach he mentioned checking the MRO, so I just went with that rather than looking at alternative approaches.

Reply:
This doesn't need to block the PR, but I like `issubclass` because I think most Python users know what `issubclass` does, but (probably?) don't know what `mro()` does.

Reply:
I agree: `issubclass` is a cleaner implementation.  I believe the class is sufficiently constructed at this point that it should work.